gluster_exporter/metric_brick.go

```diff
@@ -0,0 +1,168 @@
+)
+
+var (
+	brick_labels = []string{
+		"host",
+		"id",
+		"brick_path",
+		"volume",
+	}
+
+	glusterBrickUsedCapacity = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "brick_used_capacity",
+			Help:      "Used capacity of gluster bricks",
+		},
+		brick_labels,
+	)
+
+	glusterBrickFreeCapacity = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "brick_free_capacity",
+			Help:      "Free capacity of gluster bricks",
+		},
+		brick_labels,
+	)
+
+	glusterBrickAllCapacity = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "gluster",
+			Name:      "brick_all_capacity",
+			Help:      "Total capacity of gluster bricks",
+		},
+		brick_labels,
+	)
+)
+
+func GetGlusterBrickLabels(brick GlusterBrick) prometheus.Labels {
+	return prometheus.Labels{
+		"host": brick.Host,
+		"id": brick.Id,
+		"brick_path": brick.Path,
+		"volume": brick.VolumeName,
+	}
+}
+
+func GetLocalIP() string {
+	ifaces, _ := net.Interfaces()
+	for _, i := range ifaces {
+		addrs, _ := i.Addrs()
+		for _, addr := range addrs {
+			switch v := addr.(type) {
+			case *net.IPNet:
+				return v.IP.String()
+			case *net.IPAddr:
+				return v.IP.String()
+			}
+		}
+	}
+	return ""
+}
+
+type DiskStatus struct {
+	All  float64 `json:"all"`
+	Used float64 `json:"used"`
+	Free float64 `json:"free"`
+}
+
+type GlusterBrick struct {
+	Host       string `json:"host"`
+	Id         string `json:"id"`
+	Path       string `json:"path"`
+	PeerId     string `json:"peer-id"`
+	Type       string `json:"type"`
+	VolumeId   string `json:"volume-id"`
+	VolumeName string `json:"volume-name"`
+}
+
+type GlusterSubVolume struct {
+	ArbiterCount  uint `json:"arbiter-count"`
+	Bricks        []GlusterBrick `json:"bricks"`
+	DisperseCount uint `json:"disperse-count"`
+	Name          string `json:"name"`
+	ReplicaCount  uint `json:"replica-count"`
+	Type          string `json:"type"`
+}
+
+type GlusterVolume struct {
+	DistributeCount uint `json:"distribute-count"`
+	Id              string `json:"id"`
+	Metadata        map[string]string `json:"metadata"`
+	Name            string `json:"name"`
+	Options         map[string]string `json:"options"`
+	ReplicCount     uint `json:"replica-count"`
+	SnapList        []string `json:"snap-list"`
+	State           string `json:"state"`
+	SubVolumes      []GlusterSubVolume `json:"subvols"`
+	Transport       string `json:"transport"`
+	Type            string `json:"type"`
+}
+
+func DiskUsage(path string) (disk DiskStatus) {
+	fs := syscall.Statfs_t{}
+	err := syscall.Statfs(path, &fs)
+	if err != nil {
+		return
+	}
+	disk.All = float64(fs.Blocks * uint64(fs.Bsize))
+	disk.Free = float64(fs.Bfree * uint64(fs.Bsize))
+	disk.Used = disk.All - disk.Free
+	return
+}
+
+func BrickUtilization() {
+	fqdn, _ := os.Hostname()
+	ip, _ := net.LookupIP(fqdn)
+	//ip := GetLocalIP()
+	url := fmt.Sprintf("http://%s:24007/v1/volumes", ip)
```

Comment:
I suggest to read Volume info from a JSON file. Another thread can generate JSON from glusterd1/glusterd2 based on configuration.

- If glusterd1 => convert glusterd1 volume info to json/yml format with subvolume information similar to glusterd2
- If glusterd2 => Save JSON output of glusterd2 directly

Reply:
That's a good suggestion. Getting the json could be part of this collector itself maybe as first step.
How do we make out if its gd2 or gd1?

Reply:
config or CLI flag similar to `port` and other things

Reply:
Added a command line arg `--volinfo` to receive volinfo json file path.